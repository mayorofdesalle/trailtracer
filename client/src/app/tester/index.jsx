import '@testing-library/jest-dom';

import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';

import './mocks/i18n.mock';
import './mocks/matchMedia.mock';
import './mocks/resizeObserver.mock';

import store from '@app/store';
import GlobalStyle from '@components/style/GlobalStyle';
import Theme from '@components/style/Theme';

const Providers = ({ children }) => {
    return (
        <Provider store={store}>
            <Theme>
                <GlobalStyle />
                <BrowserRouter>
                    {children}
                </BrowserRouter>
            </Theme>
        </Provider>
    );
};

const customRender = (ui, options) =>
    render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';
export * from 'vitest';

// override render method
export { customRender as render };
