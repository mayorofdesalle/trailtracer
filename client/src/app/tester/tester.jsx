import { render } from '@testing-library/react'
import { Provider } from 'react-redux';

import './mocks/matchMedia.mock';
import store from '@app/store';
import Theme from '@components/style/Theme'

const Providers = ({ children }) => {
    return (
        <Provider store={store}>
            <Theme>
                {children}
            </Theme>
        </Provider>
    );
};

const customRender = (ui, options) =>
    render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';
export * from 'vitest';

// override render method
export { customRender as render };