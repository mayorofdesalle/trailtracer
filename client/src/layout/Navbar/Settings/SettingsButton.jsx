import { useCallback, useEffect, useRef, useState } from 'react';

import NavbarButton from '../NavbarButton';
import SettingsDropdown from './SettingsDropdown';

/**
 * SettingsButton
 * @description
 * Button to open the settings dropdown.
 **/
const SettingsButton = () => {
    const dropdown = useRef();
    const [isOpen, setIsOpen] = useState(false);

    const toggleDropdown = useCallback((e) => {
        if (isOpen) !(e.relatedTarget?.name === 'settingsButton' || dropdown.current.contains(e.relatedTarget)) && setIsOpen(false);
        else setIsOpen(true);
    }, [isOpen]);

    useEffect(() => { isOpen && dropdown.current.focus(); }, [isOpen]);

    return (
        <>
            <NavbarButton name='settingsButton' icon='equalizer' onClick={toggleDropdown} />
            {isOpen && <SettingsDropdown ref={dropdown} close={toggleDropdown} />}
        </>
    );
};

export default SettingsButton;