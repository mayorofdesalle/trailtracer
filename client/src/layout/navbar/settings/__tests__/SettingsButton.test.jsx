import { describe, expect, it, render, screen } from '@app/tester';

import SettingsButton from '../SettingsButton';

describe('SettingsButton', () => {

    it('should render correctly', () => {
        render(<SettingsButton />);
        const button = screen.getByRole('button');
        expect(button).toBeInTheDocument();

        expect(button.firstChild).toBeInTheDocument();
        expect(button.firstChild).toHaveRole('img');
    });

});