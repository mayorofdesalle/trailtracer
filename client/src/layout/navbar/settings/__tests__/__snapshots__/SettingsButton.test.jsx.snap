// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SettingsButton > should match snapshot 1`] = `
<DocumentFragment>
  <button
    class="sc-beySPh sc-kFCroH zbkUW kgqzMY"
    name="settingsButton"
    type="button"
  >
    <svg
      class="sc-eDLKkx kBnHHP"
      fill="currentColor"
      role="img"
    >
      <title>
        equalizer-line
      </title>
      <use
        href="http://localhost:3000/src/assets/icons.svg#equalizer-line"
      />
    </svg>
  </button>
</DocumentFragment>
`;
